evap/staff/templates/staff_course_comments_quick.html

```diff
@@ -0,0 +1,403 @@
+                <span class="badge badge-pill badge-secondary" data-counter="unreviewed-right"></span>
+            </div>
+        </div>
+        <div class="card-footer d-flex justify-content-center">
+            <div data-action-set="reviewing">
+                <button role="button" data-action="publish" class="btn btn-sm btn-outline-secondary">
+                    {% trans 'yes' %}
+                </button>
+                <button role="button" data-action="make_private" class="btn btn-sm btn-outline-secondary" disabled=""
+                    title="{% trans 'This text answer is about the course in general and can not refer to a single contributor' %}" data-toggle="tooltip">
+                    {% trans 'private' %}
+                </button>
+                <button role="button" data-action="hide" class="btn btn-sm btn-outline-secondary">
+                    {% trans 'no' %}
+                </button>
+            </div>
+            <div data-action-set="repeat" class="d-none">
+            {% if next_course %}
+                <a href="{% url 'staff:course_comments' next_course.semester.id next_course.id %}" data-action="next-course" class="btn btn-sm btn-primary">
+                    {% trans 'Review next course' %}
+                </a>
+            {% endif %}
+                <button role="button" data-action="repeat-all" class="btn btn-sm btn-outline-primary">
+                    {% trans 'Show all again' %}
+                </button>
+                <button role="button" data-action="repeat-unreviewed" class="btn btn-sm btn-outline-primary">
+                    {% trans 'Show unreviewed' %}
+                </button>
+            </div>
+        </div>
+    </div>
+{% endblock %}
+
+{% block additional_javascript %}
+    <script type="text/javascript">
+        var index = 0;
+        var active = null;
+        var slider;
+        var items = [];
+
+        $(document).ready(() => {
+            slider = $(".slider");
+
+            slider.on("transitionend", ".slider-item", event => {
+                updateButtons();
+                $(event.target).removeClass("to-left to-right")
+                    .css({top: "", height: ""});
+            });
+
+            slider.on("click", "[data-action]", event => {
+                handle($(event.target).data("action"), true);
+            });
+
+            $(document).keydown(event => {
+                let actions = {
+                    37: "slide-left",        // arrow left
+                    39: "slide-right",       // arrow right
+                    74: "publish",           // j
+                    75: "make_private",      // k
+                    76: "hide",              // l
+                     8: "unreview",          // backspace
+                    13: "next-course",       // enter
+                    77: "repeat-unreviewed", // m
+                    78: "repeat-all"         // n
+                };
```

Comment:
The problem with delegating hotkeys to their buttons can be lots of irritation of `:focus` and other stuff, at least it was like that when I begun with web dev. I know that a stringly typed interface is not great, but I can't think how I can avoid them on the buttons (currently `data-action` is an interface with strings, so it could be replaced with `data-action` for review, `data-slide` and `data-repeat` but it would cause more binding code imo. At least there has to be some unique identifier in the template)

Reply:
of course the buttons would need an ID which is a string, yes, but there is no problem with that or is there?

i think this solution could be much nicer if it works. can you try it please?

Reply:
Well, did my research and hacking, and I have no nice solution yet:

- Based on the lookup table for key codes to action, I can use an identifier for the button. And then `.click()` it. Problem: jQuery will fire an own jQuery event which has two sides: I can detect in the handler that the action wasn't triggered by the mouse, but from code, but the link to next course will be ignored (hence I used `.find("[data-action=next-course]")[0]` to cast it in a DOM object). It would need an extra case for the link
- I can use a single bind handler for all `[data-action]` objects which will look a lot like `handle`(kind `[data-action=#{action}]`) now or split it in three handlers for review ,  or a tuple for type and subaction (kind`[data-#{type}=#{action}]`)
- The `unreview` action has no matching button, it is the toggle state of the review buttons

Reply:
why do you need to detect whether the event was fired by the mouse or not? an extra case for the link alone is not a show stopper i'd say.

the unreview hotkey would use a selector that does not select a specific id but rather the currently selected button. or does that not work?

Reply:
I need to difference between mouse and keyboard as the review buttons have toggle behaviour with mouse but the hotkeys have do-only-one-thing behaviour. (Sorry if I'm repeating myself there) With the same reason, I can't delegate the `unreview` action to an active button as a) the code has no chance to decide whether the key was activated by keyboard with a review or a unreview or b) we had a second extra case for the unreview.

Speaking of mental models, you want to have the hotkeys imitate the mouse. This works for sliding, repeating and forwarding, but with reviewing we have a model with slightly different behaviour. I think this is the cause why I have problems translating this into code. I can hack together that follows this model, but I have problems with reading the flow.

Reply:
maybe you can do a similar thing as for unreview for the three buttons? such that their selector selects the corresponding id *and* on the activated attribute if there is any such thing. this way, the selector would find nothing if the button is pressed already.